### PR TITLE
Show keyboard API errors in UI to make troubleshooting easier

### PR DESCRIPTION
**Pull request by ValdisThomann**
No awaits were harmed in the making of this PR.

### CODE REVIEW

File: src/Routes.tsx [hunk 1]
```diff
@@ -39,7 +40,9 @@ export default () => {
         <GlobalStyle />
         {hasHIDSupport && <UnconnectedGlobalMenu />}
         <CanvasRouter />
-        <Home hasHIDSupport={hasHIDSupport}>{RouteComponents}</Home>
+        <Home hasHIDSupport={hasHIDSupport}>
+          <Route component={Errors} path={'/errors'} /> {RouteComponents}
```
Comment (olivia) [inline src/Routes.tsx:44]: Why not include this is in RouteComponents?

File: src/utils/keyboard-api.ts [hunk 2]
```diff
@@ -630,7 +632,27 @@ export class KeyboardAPI {
         'Bad Resp:',
         buffer,
       );
-      throw new Error('Receiving incorrect response for command');
+
+      const error = new Error('Receiving incorrect response for command');
+      const {message, stack} = error;
+      const {path, productId, vendorId, ...hid} = this.getHID();
+
+      store.dispatch(
```
Comment (olivia) [inline src/utils/keyboard-api.ts:640]: Firing actions for each error event seems potentially slow, could we write to a buffer and then denounce a flush?

File: src/components/panes/errors.tsx [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+import {Pane} from './pane';
+
+const Container = styled.div`
+  display: flex;
+  flex-direction: column;
+  padding: 16px;
+  border-bottom: 1px solid var(--color_accent);
+`;
+
+const KeyboardAPIErrors = () => {
```
Comment (olivia) [inline src/components/panes/errors.tsx:22]: It would be cool if we could download an error log.

File: src/Routes.tsx [hunk 4]
```diff
@@ -39,7 +40,9 @@ export default () => {
         <GlobalStyle />
         {hasHIDSupport && <UnconnectedGlobalMenu />}
         <CanvasRouter />
-        <Home hasHIDSupport={hasHIDSupport}>{RouteComponents}</Home>
+        <Home hasHIDSupport={hasHIDSupport}>
```
Comment (olivia) [inline src/Routes.tsx:43]: I can't recall where we're using an ErrorBoundary but we should redirect to the errors route if that ends up happening.

File: src/utils/keyboard-api.ts (hunk 2 above)
Comment (ValdisThomann) [reply src/utils/keyboard-api.ts:640]: It's quick enough - flip the boolean for entering this block and you can see. 

I'm not sure it's worth trying to optimise to this. Exception handling is always more expensive, but at that stage you have bigger problems on your hands than performance.

File: src/Routes.tsx (hunk 1 above)
Comment (ValdisThomann) [reply src/Routes.tsx:44]: Good idea!

File: src/components/panes/errors.tsx (hunk 3 above)
Comment (ValdisThomann) [reply src/components/panes/errors.tsx:22]: In what format do you reckon?

File: src/components/menus/global.tsx [hunk 5]
```diff
@@ -81,6 +84,17 @@ export const UnconnectedGlobalMenu = () => {
   return (
     <React.Fragment>
       <GlobalContainer>
+        {keyboardAPIErrors.length ? (
```
Comment (olivia) [inline src/components/menus/global.tsx:87]: Moving this logic into a separate component that conditionally render would be a cleaner way to hide/show this link.

File: src/components/panes/errors.tsx (hunk 3 above)
Comment (olivia) [reply src/components/panes/errors.tsx:22]: csv, tsv?

**Comment by olivia** [comment]
image

Is something supposed to show up in "response bytes" for those commands, it looks like it's mising?

**Comment by ValdisThomann** [comment]
> image
> 
> Is something supposed to show up in "response bytes" for those commands, it looks like it's mising?

Check the matching console log. For some reason we sometimes get an empty response. When the console has response bytes, this thing has response bytes

File: src/components/menus/global.tsx (hunk 5 above)
Comment (ValdisThomann) [reply src/components/menus/global.tsx:87]: Good call

File: src/components/panes/errors.tsx (hunk 3 above)
Comment (ValdisThomann) [reply src/components/panes/errors.tsx]: done!

**Review by wilba — APPROVED** [review]
I approve of this.

**Pull request merged**